server/src/main/java/io/crate/execution/engine/aggregation/impl/average/numeric/NumericAverageAggregation.java

```diff
@@ -0,0 +1,339 @@
+            mod.register(
+                Signature.aggregate(
+                    functionName,
+                    DataTypes.NUMERIC.getTypeSignature(),
+                    DataTypes.NUMERIC.getTypeSignature()),
+                NumericAverageAggregation::new
+            );
+        }
+    }
+
+    private final Signature signature;
+    private final Signature boundSignature;
+    private final DataType<BigDecimal> returnType;
+
+    @VisibleForTesting
+    private NumericAverageAggregation(Signature signature, Signature boundSignature) {
+        this.signature = signature;
+        this.boundSignature = boundSignature;
+
+        // We want to preserve the scale and precision from the
+        // numeric argument type for the return type. So we use
+        // the incoming numeric type as return type instead of
+        // the return type from the signature `avg(count::numeric(16, 2))`
+        // should return the type `numeric(16, 2)` not `numeric`
+        var argumentType = boundSignature.getArgumentDataTypes().get(0);
+        assert argumentType.id() == DataTypes.NUMERIC.id();
+        //noinspection unchecked
+        this.returnType = (DataType<BigDecimal>) argumentType;
+    }
+
+    @Nullable
+    @Override
+    public NumericAverageState newState(RamAccounting ramAccounting,
+                               Version indexVersionCreated,
+                               Version minNodeInCluster,
+                               MemoryManager memoryManager) {
+        ramAccounting.addBytes(INIT_SIZE);
+        return new NumericAverageState(new BigDecimalValueWrapper(BigDecimal.ZERO), 0L);
+    }
+
+    @Override
+    public NumericAverageState iterate(RamAccounting ramAccounting,
+                              MemoryManager memoryManager,
+                              NumericAverageState state,
+                              Input[] args) throws CircuitBreakingException {
+        if (state != null) {
+            BigDecimal value = returnType.implicitCast(args[0].value());
+            if (value != null) {
+                BigDecimal newValue = state.sum.value().add(value);
+                ramAccounting.addBytes(NumericType.sizeDiff(newValue, state.sum.value()));
+                ((BigDecimalValueWrapper) state.sum).setValue(newValue);
+                state.count++;
+            }
+        }
+        return state;
+    }
+
+    @Override
+    public NumericAverageState reduce(RamAccounting ramAccounting,
+                               NumericAverageState state1,
+                               NumericAverageState state2) {
+        if (state1 == null) {
+            return state2;
+        }
+        if (state2 == null) {
+            return state1;
+        }
+        BigDecimal newValue = state1.sum.value().add(state2.sum.value());
+        ramAccounting.addBytes(NumericType.sizeDiff(newValue, state1.sum.value()));
+        ((BigDecimalValueWrapper) state1.sum).setValue(newValue);
+        state1.count += state2.count;
```

Comment:
See my previous comment on `iterate`, could be also solved by implementing e.g `NumericAverageState.merge(state)`